File: Subtask3.py
M = 9
MOD = 998244353
DP = [[0,1,0,0,0,0,0,0,0],
	  [0,0,1,0,0,0,0,0,0],
	  [0,0,0,1,0,0,0,0,0],
	  [0,0,0,0,1,0,0,0,0],
	  [0,0,0,0,0,1,0,0,0],
	  [0,0,0,0,0,0,1,0,0],
	  [1,0,0,0,0,0,0,1,0],
	  [0,0,0,0,0,0,0,0,1],
	  [1,0,0,0,0,0,0,0,0]]

def MatrixMult(A, B, MOD):
    Bx = len(B[0])
    res = [[0 for _ in range(Bx)]for _ in range(M)]
    for i in range(M):
        for j in range(Bx):
            for k in range(M):
                res[i][j] += A[i][k] * B[k][j]
                res[i][j] %= MOD
    return res

N,T = map(int,input().split())
A = list(map(int,input().split()))

Ans = [[0] for _ in range(M)]
for i in range(N):
	Ans[A[i]][0] += 1

while T:
	if T&1:
		Ans = MatrixMult(DP, Ans, MOD)
	DP = MatrixMult(DP, DP, MOD)
	T >>= 1

for i in range(9):
    print(Ans[i][0])

File: Subtask3.cpp
#include <iostream>
#include <vector>
#include <cmath>
#define int long long
using namespace std;

const int MOD = 998244353;
const int M = 9;

vector<vector<int>> DP = {
    {0, 1, 0, 0, 0, 0, 0, 0, 0},
    {0, 0, 1, 0, 0, 0, 0, 0, 0},
    {0, 0, 0, 1, 0, 0, 0, 0, 0},
    {0, 0, 0, 0, 1, 0, 0, 0, 0},
    {0, 0, 0, 0, 0, 1, 0, 0, 0},
    {0, 0, 0, 0, 0, 0, 1, 0, 0},
    {1, 0, 0, 0, 0, 0, 0, 1, 0},
    {0, 0, 0, 0, 0, 0, 0, 0, 1},
    {1, 0, 0, 0, 0, 0, 0, 0, 0}
};

vector<vector<int>> matrixMult(vector<vector<int>>& A, vector<vector<int>>& B) {
    int Bx = B.size();

    vector<vector<int>> result(M, vector<int>(Bx, 0));

    for (int i = 0; i < M; ++i) { 			// A_row
        for (int j = 0; j < Bx; ++j) { 		// B_col
            for (int k = 0; k < M; ++k) { 	// A_col = B_row
                result[i][j] += (A[i][k] % MOD * B[k][j] % MOD) % MOD;
				result[i][j] %= MOD;
            }
        }
    }

    return result;
}

signed main() {
    int N, T;
    cin >> N >> T;

	// Ans_0 (= V)
    vector<vector<int>> Ans(M, vector<int>(1,0));
    for (int i = 0; i < N; ++i) {
        int A; cin >> A;
        Ans[A][0] += 1;
    }

    while (T) {
        if (T & 1) {
            Ans = matrixMult(DP, Ans);
        }
        DP = matrixMult(DP, DP);
        T >>= 1;
    }

    for (int i = 0; i < 9; ++i) {
        cout << Ans[i][0] << endl;
    }

    return 0;
}
